rts/lra/test/crash/src/test/java/io/narayana/lra/arquillian/LRACoordinatorRecoveryIT.java

```diff
@@ -0,0 +1,381 @@
+        clearRecoveryLogFromObjectStore();
+        clearRecoveryLogFromFS();
+    }
+
+    @After
+    public void after() {
+        if (client != null) {
+            client.close();
+        }
+
+        // Cleans Object Store in case there are still active transactions
+        clearRecoveryLogFromObjectStore();
+        clearRecoveryLogFromFS();
+
+        for (Map.Entry<String, String> entry : containerDeploymentMap.entrySet()) {
+            stopContainer(entry.getKey(), entry.getValue());
+        }
+    }
+
+    /**
+     * Arquillian uses this method to deploy lra-participant to the container targeted with @TargetsContainer.
+     * Together with the extension LRACoordinatorExtension, a Server-Client environment is created with two
+     * Wildfly containers, hosting lra-coordinator and lra-participant respectively.
+     */
+    @Deployment(name = LRA_PARTICIPANT_DEPLOYMENT_QUALIFIER, testable = false, managed = false)
+    @TargetsContainer("lra-participant-server")
+    public static WebArchive deploy() {
+        return Deployer.createDeployment(LRA_PARTICIPANT_DEPLOYMENT_QUALIFIER, LRAListener.class);
+    }
+
+    @Test
+    public void lraCoordinatorShortTimeoutLRA(@ArquillianResource @OperateOnDeployment(LRA_PARTICIPANT_DEPLOYMENT_QUALIFIER) URL baseURL)
+            throws URISyntaxException {
+
+        String lraId;
+        URI lraListenerURI = UriBuilder.fromUri(baseURL.toURI()).path(LRAListener.LRA_LISTENER_PATH).build();
+
+        // Starts an LRA with a short time limit by invoking a resource annotated with @LRA. The time limit is
+        // defined as LRAListener.LRA_SHORT_TIMELIMIT
+        Response response = client
+                .target(lraListenerURI)
+                .path(LRAListener.LRA_LISTENER_ACTION)
+                .request()
+                .put(null);
+
+        Assert.assertEquals("LRA participant action", 200, response.getStatus());
+
+        // Restarts lra-coordinator to simulate a crash
+        restartContainer(LRA_COORDINATOR_CONTAINER_QUALIFIER);
+
+        // Waits for a period of time longer than the timeout of the LRA Transaction
+        doWait((LRAListener.LRA_SHORT_TIMELIMIT + 1L) * 1000);
+
+        // The LRA transaction could have been started via lraClient but it is useful to test the filters as well
+        lraId = getFirstLRAFromFS();
+        assertNotNull("A new LRA should have been added to the object store before the JVM was halted.", lraId);
+        lraId = String.format("%s/%s", lraClient.getCoordinatorUrl(), lraId);
+
+        // Waits for a period of time longer than the timeout of the LRA Transaction
+        doWait((LRAListener.LRA_SHORT_TIMELIMIT + 1L) * 1000);
+
+        // Checks recovery
+        LRAStatus status = getStatus(new URI(lraId));
+
+        LRALogger.logger.infof("%s: Status after restart is %s%n", status == null ? "GONE" : status.name());
+
+        if (status == null || status == LRAStatus.Cancelling) {
+            int sc = recover();
```

Comment:
@mmusgrov this if for you ;-)

Reply:
If you take a look at what the recover call returns you should understand what it means, the method is not very complicated. That applies to both of you.

But the detail is that it would once have returned a JAX-RS status code and has subsequently been changed to return the number of LRAs that are in need of recovery. Now if you want to know the history of why that change was made then it probably involves a refactoring of the code base, I think you, martin, were involved in some of that work.

Reply:
it doesn't matter that much and I must admit I don't have time to open every PR in IDE :) We can revisit it later if needed.